### ISSUE TITLE
How to obtain the Correct Outputs of the Decoders?

### ISSUE THREAD

**Issue by jessxphil** [deezer/linear_graph_autoencoders#3]
I'm running the GAE algorithm using the 'Cora' graph dataset. 

But the issue is, the size of my input Encoder Adjacency, A, is double the size of my output Adjacency, Ahat, Decoder output shape. So, somewhere along this process I think I'm losing data and I don't know why. I'm running the algorithm without any changes to the source code other than output shapes/size, etc. 

`!python3 "/content/linear_gae/train.py" --model=linear_vae --dataset=cora --task=link_prediction 
`

output: 
```
Input Data Adjacency size and shape, 2708 (2708, 1)
Encoded data Embedding Space (embs) size and shape, 43328 (2708, 16)
Decoded data Adjacency (preds_all) size and shape, 1054 (1054, 1)
```

I simply want to embed a list of graphs, and decode those graphs. The decoder should attempt to "predict" the exact same graphs (# of nodes and # of graphs should remain the same) that were embedded. Please confirm my intuition. 

_In other words:_ 
I expect the same number of graphs that go into the Encoder should equal the same number of graphs leaving the Decoder. The number of nodes per graph should remain the same with the links changing due to Link Prediction. Please confirm.

**Comment by GuillaumeSalhaGalvan** [comment]
Dear @jessxphil,

Thank you for your interest!

I assume you are referring to the `preds_all` from `evaluation.py`, which does not include the entire matrix _\hat{A}_. It only gathers predicted values _\hat{A}\_{ij}_ for node pairs _(i,j)_ of the link prediction test set, i.e. the edges masked during training and the randomly picked unconnected node pairs (see our [Section 4.1. ](https://arxiv.org/pdf/2001.07614.pdf) on experimental evaluation).


> The decoder should attempt to "predict" the exact same graphs (# of nodes and # of graphs should remain the same) that were embedded. Please confirm my intuition.

Indeed! In order to reconstruct the **entire** 2708x2708  _\hat{A}_, for Cora, you can instead call `model.reconstructions` together during a correct reshaping.  Specifically, after training your model in `train.py`, you can add the following commands:

``` 
# Retrieving all inner products zi^T zj
adj_hat = sess.run(model.reconstructions, feed_dict = feed_dict)
# Reshaping matrix
adj_hat = np.reshape(adj_hat, adj.shape)
#Adding an element-wise sigmoid to normalize elements in the [0,1] set
adj_hat = 1 / (1 + np.exp(-adj_hat))
```

It should give you what you are looking for:

```
print(adj_hat.shape)
(2708, 2708)
```

```
print(adj_hat[0:5,0:5])
[[0.9638505  0.3758499  0.47716868 0.74183184 0.15876555]
 [0.3758499  0.984312   0.96703297 0.38611087 0.7850153 ]
 [0.47716868 0.96703297 0.98263264 0.43402663 0.6424965 ]
 [0.74183184 0.38611087 0.43402663 0.82597256 0.4580344 ]
 [0.15876555 0.7850153  0.6424965  0.4580344  0.9594626 ]]
```

Best,

Guillaume

**Comment by jessxphil** [comment]
*/dramatically drops to knees and weeps/* Thank you so much @GuillaumeSalhaGalvan! I was able to extract the expected data using the sample dataset, Cora. 

Now, */cracks fingers/*, I'll be using your work to run my own data and this is where I expect issues. Why? The example datasets are networks graphs. My data includes a large list (10k+) of small graphs of size (33 x 33) for each graph. 

I was looking for a simple structure to: 
  1. Encode my list of graphs into an Embedding Space
  2. Perform Unsupervised Node Clustering on said Embedding Space based on learned features/represenataion of my data. 
  3. Decode my embedding space and use your reconstruction model to output my initial list of graphs (w/ loss, etc.). 

Sigh... Is that possible?

**Comment by GuillaumeSalhaGalvan** [comment]
Nice! I am now closing this issue, but feel free to send me an e-mail if you encounter some problem.

Regarding your project: what is your ultimate goal?
- Are you aiming at clustering the 10k+ _graphs_ themselves? If yes, you might try to look for alternative methods, that instead learn _graph-level representations_. In such a setting, one z_i vector corresponds to one graph _i_ from a set of several graphs. Then, you can cluster these z_i vectors, e.g. via k-means, to obtain clusters of graphs.
- Are you aiming at clustering the _nodes_ from all these graphs? If yes, I guess the easiest* way would be to assemble all your small graphs into _one large graph_ with a lot of isolated subgraphs components, and to subsequently train a GAE on this large graph.

*"The easiest way" is not necessarily "the optimal way" however :) I am not sure to which extent this strategy would work (and I would need to know more about your project/data to have a more informative opinion on this).
